### PR TITLE
QgsSettingsEntry implementations rely on templated base class ; new option: former values for settings

### PR DESCRIPTION
**Pull request by 3nids**
This PR brings a few things at once:

* QgsSettingsEntry implementations were refactored using templates to avoid code duplication
* Settings have now options, the first available allows to save the former value of a setting (with test)
* QgsSnappingConfig type flag and config mode enum are moved to Qgis 
* improve Qgis enum/flag value-key conversions to determine if the conversion suceeded (with test)
* remember former capture technique and use it to quickly toggle techniques

this brings a better fix for the toggling of digitzing techniques (not so well fixed in #47480)

### CODE REVIEW

**Comment by 3nids** [comment]
Unrelated test failure (mingw)

File: src/core/qgis.h [hunk 1]
```diff
@@ -200,6 +200,35 @@ class CORE_EXPORT Qgis
...
+    /**
+     * Settings options
+     * \since QGIS 3.26
+     */
+    enum class SettingsOption : int
+    {
+      SaveFormerValue = 1 << 1, //<! Save the former value of the settings
+    };
+    Q_ENUM( SettingsOption )
+    Q_DECLARE_FLAGS( SettingsOptions, SettingsOption )
```
Comment (nyalldawson) [inline src/core/qgis.h:229]: don't forget the Q_DECLARE_OPERATORS_FOR_FLAGS

File: tests/src/python/test_qgssettingsentry.py [hunk 2]
```diff
@@ -72,7 +72,7 @@ def test_settings_entry_base(self):
         self.assertEqual(settingsEntryVariant.valueAsVariant(), 43)
 
         # Settings type
-        self.assertEqual(settingsEntryVariant.settingsType(), QgsSettingsEntryBase.SettingsType.Variant)
+        self.assertEqual(settingsEntryVariant.settingsType(), Qgis.SettingsType.Variant)
```
Comment (nyalldawson) [inline tests/src/python/test_qgssettingsentry.py:75]: is this change just to remove deprecation warnings? Or to get the tests to pass?

File: src/app/options/qgsoptions.cpp [hunk 3]
```diff
@@ -1161,19 +1161,10 @@ QgsOptions::QgsOptions( QWidget *parent, Qt::WindowFlags fl, const QList<QgsOpti
...
-          Qgis::SnappingType::Segment,
-          Qgis::SnappingType::Centroid,
-          Qgis::SnappingType::MiddleOfSegment,
-          Qgis::SnappingType::LineEndpoint,
-          Qgis::SnappingType::Area,
-        } )
-  {
-    mDefaultSnapTypeComboBox->addItem( QgsSnappingConfig::snappingTypeFlagToIcon( type ),
-                                       QgsSnappingConfig::snappingTypeFlagToString( type ),
+  for ( Qgis::SnappingType type : qgsEnumMap<Qgis::SnappingType>().keys() )
```
Comment (nyalldawson) [inline src/app/options/qgsoptions.cpp:1164]: This looks rather inefficient -- we build a map of the key to values, then throw away the values and build another list of just the keys. This calls for a more appropriate convience method to just return a list/vector of the keys directly, without the multiple container constructions and iterations.

**Comment by domi4484** [comment]
Just a detail, would it be possible to rename the file `qgssettingsentry` to `qgssettingsentrybase` and `qgssettingsentryimpl` back to `qgssettingsentry` to keep using the same include overall in code?

**Review by DelazJ — COMMENTED** [review]
Some typos and punctuation suggestions. Didn't comment all since there are many occurrences of same sentences you can easily fix elsewhere than from github suggestions

File: src/core/qgis.h (hunk 1 above)
Comment (3nids) [reply src/core/qgis.h:229]: this is done yes

File: tests/src/python/test_qgssettingsentry.py (hunk 2 above)
Comment (3nids) [reply tests/src/python/test_qgssettingsentry.py:75]: I used the sip patching technique technique and did a global replace. So there is no API break.

File: src/app/options/qgsoptions.cpp (hunk 3 above)
Comment (3nids) [reply src/app/options/qgsoptions.cpp]: I just added a new method which does this [f2243e1](https://github.com/qgis/QGIS/pull/47535/commits/f2243e18b8c7a35156d2e7012bbcfbda8a4b3e4b)

File: src/core/settings/qgssettingsentryimpl.h [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+     * Constructor for QgsSettingsEntryString.
+     *
+     * The \a key argument specifies the final part of the settings key.
+     * The \a section argument specifies the section.
+     * The \a defaultValue argument specifies the default value for the settings entry.
+     * The \a description argument specifies a description for the settings entry.
+     * The \a options arguments specifies the options for the settings entry.
+     * The \a minLength argument specifies the minimal length of the string value. 0 means no limit.
+     * The \a maxLength argument specifies the maximal length of the string value. -1 means no limit.
+     * By -1 the there is no limit
```
[the remark this answers is not in the document]
Comment (3nids) [reply src/core/settings/qgssettingsentryimpl.h:105]: fixed, thanks

**Comment by 3nids** [comment]
> Just a detail, would it be possible to rename the file qgssettingsentry to qgssettingsentrybase and qgssettingsentryimpl back to qgssettingsentry to keep using the same include overall in code?

I'm -0 here. It's quite aligned with the rest of the code base. I'd keep qgssettingsentryimpl as is. But I could rename qgsettingsentry to qgssettingsentrybase if you think it's a good idea.

File: src/core/qgis.h [hunk 5]
```diff
@@ -1763,27 +1824,38 @@ template<class T> const QMap<T, QString> qgsEnumMap() SIP_SKIP
 
 /**
  * Returns the value for the given key of an enum.
+ * If \a returnOk is given, it defines if the value could be converted to the key
```
[the remark this answers is not in the document]
Comment (3nids) [reply src/core/qgis.h:1842]: is there a policy about introduction of parameters in the documentation? I am a bit concerned about the noise.

[the remark this answers is not in the document]
Comment (domi4484) [reply src/core/qgis.h:1842]: I don't now about a policy, I have just seen it at other places. I think this is an interesting information when writing plugins.

**Comment by domi4484** [comment]
> > Just a detail, would it be possible to rename the file qgssettingsentry to qgssettingsentrybase and qgssettingsentryimpl back to qgssettingsentry to keep using the same include overall in code?
> 
> I'm -0 here. It's quite aligned with the rest of the code base. I'd keep qgssettingsentryimpl as is. But I could rename qgsettingsentry to qgssettingsentrybase if you think it's a good idea.

Ok, I see, no you can keep qgssettingsentry for me

File: src/core/qgis.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (3nids) [reply src/core/qgis.h:1842]: that would makes sense for plugins indeed.
This method is cpp only, so I'd keep it as. Thanks.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**